### PR TITLE
Make location data structured

### PR DESCRIPTION
**Pull request by Schparky**
Replaces destination and origin fields with objects to include political (country, etc.) and geo (lat/long) data. Breaks compatibility with prior GraphQL schema.

### CODE REVIEW

**Comment by Schparky** [comment]
Redesigning the database implementation to use a new table for the location information. Will resubmit when that is done.

File: application/gqlgen/user_test.go [hunk 1]
```diff
@@ -117,41 +116,82 @@ func Fixtures_UserQuery(as *ActionSuite, t *testing.T) UserQueryFixtures {
...
 	}
 
-	TestUser = userFixtures[0]
-	TestUser.AdminRole = nulls.NewString(domain.AdminRoleSuperDuperAdmin)
-	c.MustPost(query, &usersResp)
+	var resp UserResponse
 
-	if err := as.DB.Load(&(userFixtures[1]), "PhotoFile"); err != nil {
-		t.Errorf("failed to load user fixture, %s", err)
+	allFields := `{ id email nickname adminRole photoURL posts (role: CREATEDBY) {id} organizations {id} location {description country latitude longitude} }`
```
Comment (Baggerone) [inline application/gqlgen/user_test.go]:
```suggestion
	allFields := `{ id email nickname adminRole photoURL posts (role: CREATEDBY) {id} organizations {id} ` + 
        `location {description country latitude longitude} }`
```

Comment (Baggerone) [reply application/gqlgen/user_test.go]: (To keep it visible in the github review window.)

File: application/models/post.go [hunk 2]
```diff
@@ -292,3 +294,55 @@ func (p *Post) FindByUserAndUUID(ctx context.Context, user User, uuid string, se
...
+
+	return &location, nil
+}
+
+func (p *Post) SetDestination(location Location) error {
+	if p.DestinationID.Valid {
+		location.ID = p.DestinationID.Int
+		p.Destination = location
+		return DB.Update(&p.Destination)
+	} else {
```
Comment (Baggerone) [inline application/models/post.go]: Maybe remove the `else` line (here and in SetOrigin)?

File: application/models/post.go [hunk 3]
```diff
@@ -292,3 +294,55 @@ func (p *Post) FindByUserAndUUID(ctx context.Context, user User, uuid string, se
...
+	}
+	location := Location{}
+	if err := DB.Find(&location, p.OriginID); err != nil {
+		return nil, err
+	}
+
+	return &location, nil
+}
+
+func (p *Post) SetDestination(location Location) error {
```
Comment (Baggerone) [inline application/models/post.go]: I could use doc comments on this method and the SetOrigin method.

File: application/models/post.go [hunk 4]
```diff
@@ -292,3 +294,55 @@ func (p *Post) FindByUserAndUUID(ctx context.Context, user User, uuid string, se
...
+	if p.DestinationID.Valid {
+		location.ID = p.DestinationID.Int
+		p.Destination = location
+		return DB.Update(&p.Destination)
+	} else {
+		if err := DB.Create(&location); err != nil {
+			return err
+		}
+		p.DestinationID = nulls.NewInt(location.ID)
```
Comment (Baggerone) [inline application/models/post.go]: Maybe add a test for each of these methods that touch this part of the code.  I'm wondering if the location will have its ID already at this point.

File: application/models/user.go [hunk 5]
```diff
@@ -383,3 +386,29 @@ func (u *User) uniquifyNickname() error {
...
+
+	return &location, nil
+}
+
+func (u *User) SetLocation(location Location) error {
+	if u.LocationID.Valid {
+		location.ID = u.LocationID.Int
+		u.Location = location
+		return DB.Update(&u.Location)
+	} else {
```
Comment (Baggerone) [inline application/models/user.go]: Maybe remove the `else` line?

File: application/models/user.go [hunk 6]
```diff
@@ -383,3 +386,29 @@ func (u *User) uniquifyNickname() error {
...
+	}
+	location := Location{}
+	if err := DB.Find(&location, u.LocationID); err != nil {
+		return nil, err
+	}
+
+	return &location, nil
+}
+
+func (u *User) SetLocation(location Location) error {
```
Comment (Baggerone) [inline application/models/user.go]: Doc comment and unit test?

File: application/gqlgen/models_gen.go [hunk 7]
```diff
@@ -27,6 +27,13 @@ type CreateOrganizationInput struct {
 	AuthConfig string  `json:"authConfig"`
 }
 
+type LocationInput struct {
+	Description string   `json:"description"`
+	Country     *string  `json:"country"`
```
[the remark this answers is not in the document]
Comment (Schparky) [reply application/gqlgen/models_gen.go]: It's correct by the schema. Do you think it should be required? Thinking about it now, it makes sense.

File: application/gqlgen/gqlgenutils.go [hunk 8]
```diff
@@ -19,3 +13,35 @@ func GetStringFromNullsString(inString nulls.String) *string {
...
+		*output = nulls.NewFloat64(*input)
+	}
+}
+
+func convertGqlLocationInputToDBLocation(input LocationInput) models.Location {
+	l := models.Location{
+		Description: input.Description,
+	}
+
+	setOptionalStringField(input.Country, &(l.Country))
```
[the remark this answers is not in the document]
Comment (Schparky) [reply application/gqlgen/gqlgenutils.go]: As of now, yes.

File: application/models/post.go (hunk 4 above)
Comment (Schparky) [reply application/models/post.go]: I was planning on it, but wanted to keep pace with the corresponding UI update. I'm pretty sure `Test_UpdatePost` exercises both paths. I still intend to do these unit tests, though.

**Pull request merged**